modules/gateway.go

```diff
@@ -0,0 +1,54 @@
+package modules
+
+import (
+	"github.com/NebulousLabs/Sia/consensus"
+	"github.com/NebulousLabs/Sia/network"
+)
+
+type GatewayInfo struct {
+	Peers []network.Address
+}
+
+// A Gateway facilitates the interactions between the local node and remote
+// nodes (peers). It relays incoming blocks and transactions to local modules,
+// and broadcasts outgoing blocks and transactions to peers. In a broad sense,
+// it is responsible for ensuring that the local consensus set is consistent
+// with the "network" consensus set.
+type Gateway interface {
+	// Bootstrap joins the Sia network and establishes an initial peer list.
+	Bootstrap(network.Address) error
+
+	// AddPeer adds a peer to the Gateway's peer list. The peer
+	// may be rejected. AddPeer is also an RPC.
+	AddPeer(network.Address) error
+
+	// RemovePeer removes a peer from the Gateway's peer list.
+	RemovePeer(network.Address) error
+
+	// Synchronize synchronizes the local consensus set with the sets of known
+	// peers.
+	Synchronize() error
+
+	// RelayBlock announces a block to all of the Gateway's known peers.
+	RelayBlock(consensus.Block) error
+
+	// RelayTransaction announces a transaction to all of the Gateway's
+	// known peers.
+	RelayTransaction(consensus.Transaction) error
+
+	// AddMe is the RPC version of AddPeer. It is assumed that the supplied
+	// peer is the peer making the RPC.
+	AddMe(network.Address) error
+
+	// SendBlocks is an RPC that returns a set of sequential blocks following
+	// the most recent known block ID in of the 32 IDs provided. The number of
+	// blocks returned is unspecified.
+	SendBlocks([32]consensus.BlockID) ([]consensus.Block, error)
+
+	// SharePeers is an RPC that returns a set of the Gateway's peers. The
+	// number of peers returned is unspecified.
+	SharePeers() ([]network.Address, error)
```

Comment:
hmm. I tried to look into how Bitcoin handles this, but it wasn't clear if they protect against this attack either. The main difference is that there is no explicit "AddMe" call; instead, nodes you connect to choose whether or not they want to add you. But it's not much different -- instead of calling "AddMe", you call "ver".
Bitcoin also periodically removes addresses that it hasn't heard from in a while, but the spec says this only occurs if it has more than 1000 addresses. So I dunno.